consensus/polybft/checkpoint_manager.go

```diff
@@ -0,0 +1,241 @@
+		"bytes32 eventRoot, tuple(address _address, uint256[4] blsKey)[] nextValidators" + ")")
+
+	// frequency at which checkpoints are sent to the rootchain (in blocks count)
+	defaultCheckpointsOffset = uint64(900)
+)
+
+// checkpointManager encapsulates logic for checkpoint data submission
+type checkpointManager struct {
+	// sender address
+	sender types.Address
+	// blockchain is abstraction for blockchain
+	blockchain blockchainBackend
+	// consensusBackend is abstraction for polybft consensus specific functions
+	consensusBackend polybftBackend
+	// rootchain represents abstraction for rootchain interaction
+	rootchain rootchainInteractor
+	// checkpointsOffset represents offset between checkpoint blocks (applicable only for non-epoch ending blocks)
+	checkpointsOffset uint64
+	// latestCheckpointID represents last checkpointed block number
+	latestCheckpointID uint64
+}
+
+// newCheckpointManager creates a new instance of checkpointManager
+func newCheckpointManager(sender types.Address, checkpointOffset uint64, interactor rootchainInteractor,
+	blockchain blockchainBackend, backend polybftBackend) *checkpointManager {
+	r := interactor
+	if interactor == nil {
+		r = &defaultRootchainInteractor{}
+	}
+
+	return &checkpointManager{
+		sender:            sender,
+		blockchain:        blockchain,
+		consensusBackend:  backend,
+		rootchain:         r,
+		checkpointsOffset: checkpointOffset,
+	}
+}
+
+// getCurrentCheckpointID queries CheckpointManager smart contract and retrieves current checkpoint id
+// (the latest checkpoint block number)
+func (c checkpointManager) getCurrentCheckpointID() (uint64, error) {
+	checkpointIDMethodEncoded, err := currentCheckpointIDMethod.Encode([]interface{}{})
+	if err != nil {
+		return 0, fmt.Errorf("failed to encode currentCheckpointId function parameters: %w", err)
+	}
+
+	currentCheckpointID, err := c.rootchain.Call(c.sender, helper.CheckpointManagerAddress, checkpointIDMethodEncoded)
+	if err != nil {
+		return 0, fmt.Errorf("failed to invoke currentCheckpointId function on the rootchain: %w", err)
+	}
+
+	checkpointID, err := strconv.ParseUint(currentCheckpointID, 0, 64)
+	if err != nil {
+		return 0, fmt.Errorf("failed to convert current checkpoint id '%s' to number: %w",
+			currentCheckpointID, err)
+	}
+
+	return checkpointID, nil
+}
+
+// submitCheckpoint sends a transaction which with checkpoint data to the rootchain
+func (c checkpointManager) submitCheckpoint(latestHeader types.Header, isEndOfEpoch bool) error {
+	lastCheckpointBlockNumber, err := c.getCurrentCheckpointID()
+	if err != nil {
+		return err
+	}
+
+	nonce, err := c.rootchain.GetPendingNonce(c.sender)
+	if err != nil {
+		return err
+	}
+
+	checkpointManagerAddr := ethgo.Address(helper.CheckpointManagerAddress)
+	txn := &ethgo.Transaction{
+		To: &checkpointManagerAddr,
+	}
+
+	initialBlockNumber := lastCheckpointBlockNumber + 1
+
+	var parentExtra *Extra
+
```

Comment:
```suggestion
```

Reply:
Unfortunatelly I think I've tried it, but linter complains: declarations must not be cuddled.